internal/graph/object_providers.go

```diff
@@ -0,0 +1,140 @@
+
+type simpleRecursiveObjectProvider struct {
+	mapper storage.TupleMapper
+	ts     *typesystem.TypeSystem
+	ds     storage.RelationshipTupleReader
+}
+
+func newSimpleRecursiveObjectProvider(ts *typesystem.TypeSystem, ds storage.RelationshipTupleReader) (*simpleRecursiveObjectProvider, error) {
+	if ts == nil || ds == nil {
+		return nil, fmt.Errorf("%w: nil arguments", openfgaErrors.ErrUnknown)
+	}
+	return &simpleRecursiveObjectProvider{ts: ts, ds: ds}, nil
+}
+
+var _ objectProvider = (*simpleRecursiveObjectProvider)(nil)
+
+func (s *simpleRecursiveObjectProvider) End() {
+	if s.mapper != nil {
+		s.mapper.Stop()
+	}
+}
+
+func (s *simpleRecursiveObjectProvider) Begin(ctx context.Context, req *ResolveCheckRequest) (chan usersetMessage, error) {
+	if req == nil {
+		return nil, fmt.Errorf("%w: nil request", openfgaErrors.ErrUnknown)
+	}
+	// Note: we set sortContextualTuples to false because we don't care about ordering of results,
+	// since the consumer is using hashsets to check for intersection.
+	userIter, err := checkutil.IteratorReadStartingFromUser(ctx, s.ts, s.ds, req,
+		tuple.ToObjectRelationString(tuple.GetType(req.GetTupleKey().GetObject()), req.GetTupleKey().GetRelation()),
+		nil, false)
+	if err != nil {
+		return nil, err
+	}
+	usersetFromUserIter := storage.WrapIterator(storage.ObjectIDKind, userIter)
+	s.mapper = usersetFromUserIter
+
+	// note: this function will close the channel
+	userToUsersetMessageChan := streamedLookupUsersetFromIterator(ctx, usersetFromUserIter)
+
+	return userToUsersetMessageChan, nil
+}
+
+type complexRecursiveObjectProvider struct {
+	concurrencyLimit uint32
+	wg               sync.WaitGroup
+	ts               *typesystem.TypeSystem
+}
+
+func newComplexRecursiveObjectProvider(concurrencyLimit uint32, ts *typesystem.TypeSystem) (*complexRecursiveObjectProvider, error) {
+	if ts == nil {
+		return nil, fmt.Errorf("%w: nil typesystem", openfgaErrors.ErrUnknown)
+	}
+	return &complexRecursiveObjectProvider{concurrencyLimit: concurrencyLimit, wg: sync.WaitGroup{}, ts: ts}, nil
+}
+
+var _ objectProvider = (*complexRecursiveObjectProvider)(nil)
+
+func (c *complexRecursiveObjectProvider) End() {
+	c.wg.Wait()
+}
+
+func (c *complexRecursiveObjectProvider) Begin(ctx context.Context, req *ResolveCheckRequest) (chan usersetMessage, error) {
+	if req == nil {
+		return nil, fmt.Errorf("%w: nil request", openfgaErrors.ErrUnknown)
+	}
+	objectType, relation := tuple.GetType(req.GetTupleKey().GetObject()), req.GetTupleKey().GetRelation()
+	userType := tuple.GetType(req.GetTupleKey().GetUser())
+	operands, ok := c.ts.IsRelationWithRecursiveTTUAndAlgebraicOperations(objectType, relation, userType)
+	if !ok {
+		return nil, fmt.Errorf("%w: unsupported model", openfgaErrors.ErrUnknown)
+	}
+
+	outChannel := make(chan usersetMessage, len(operands)) // buffered so that the first writes don't block on the caller reading from the channel
+	pool := concurrency.NewPool(ctx, max(int(c.concurrencyLimit), len(operands)))
```

Comment:
Could there be a case where `operands` length == 0? If so, should that be handled?

Reply:
No, it cannot happen. `IsRelationWithRecursiveTTUAndAlgebraicOperations`  has:

```go
satisfies := ttuEdgeSatisfiesCond && len(usersets) >= 1
return usersets, satisfies
```